### PR TITLE
Add Anndata peek capability

### PR DESCRIPTION
**Pull request by mtekman**
This is an attempt to add peek and display functionality to the AnnData type. Targets the `release_20.01` branch.
 It is part of the GCC 2020 collaboration fest.

TODO:

* [X] Extract relevant metadata
  * [x] Passing local upload tests
* [X] Format a peek string
  * [x] Passing local upload tests
* [ ] Display a data matrix
* [x] Passing all tests

### CODE REVIEW

**Comment by mtekman** [comment]
Question: Is it possible to access metadata elements from within the `set_peek` function?

**Comment by bgruening** [comment]
@mtekman like here? https://github.com/galaxyproject/galaxy/blob/17b0f4d5077666e78abd7af8388705eb27419a7c/lib/galaxy/datatypes/binary.py#L161

**Comment by mtekman** [comment]
I am getting some strange errors then, but I think we (@astrovsky01 and I) could solve it

**Comment by mtekman** [comment]
@bgruening I'm not sure how related this error is to my PR:

https://jenkins.galaxyproject.org/job/docker-framework-py3/4705/console

```
test/functional/test_toolbox_pytest.py::test_tool[composite_test_1] FAILED [  7%]
...
galaxy.jobs.runners ERROR 2020-08-05 16:18:32,903 (29) Failure preparing job
Traceback (most recent call last):
  File "/galaxy/lib/galaxy/util/__init__.py", line 1053, in unicodify
    value = text_type(value)
  File "/galaxy_venv/lib/python3.6/site-packages/Cheetah/Template.py", line 1053, in __unicode__
    return getattr(self, mainMethName)()
  File "cheetah_DynamicallyCompiledCheetahTemplate_1596644312_8808622_87732.py", line 89, in respond
AttributeError: 'builtin_function_or_method' object has no attribute 'extra_files_path'
```

**Comment by dannon** [comment]
@mtekman It is not related to this PR, no worries.  It's fixed in #10071

**Comment by mtekman** [comment]
@dannon Phew, thanks!

**Comment by bgruening** [comment]
Restartet the test!

File: lib/galaxy/datatypes/binary.py [hunk 1]
```diff
@@ -950,6 +977,116 @@ def sniff(self, filename):
...
+                        ', '.join(sorted(names))
+                    )
+                return ""
+
+            peekstr = "[n_obs x n_vars]\n    %d x %d" % tuple(tmp.shape)
+            peekstr += _makelayerstrings("obs", tmp.obs_count, tmp.obs_layers)
+            peekstr += _makelayerstrings("var", tmp.var_count, tmp.var_layers)
+            peekstr += _makelayerstrings("obsm", tmp.obsm_count, tmp.obsm_layers)
+            peekstr += _makelayerstrings("varm", tmp.varm_count, tmp.varm_layers)
+            peekstr += _makelayerstrings("uns", tmp.uns_count, tmp.uns_layers)
```
Comment (bgruening) [inline lib/galaxy/datatypes/binary.py:1076]: Is X here missing?

**Comment by mtekman** [comment]
X is the main matrix, where we only need to know the current shape of it.
This we take care of already above

On Saturday, 8 August 2020, Björn Grüning <notifications@github.com> wrote:
> @bgruening commented on this pull request.
>
> ________________________________
>
> In lib/galaxy/datatypes/binary.py:
>
>> + "Format the layers."
> + if layer in tmp.layers_names:
> + return "\n[%s]: %d %s\n %s" % (
> + layer,
> + count,
> + "layer" if count == 1 else "layers",
> + ', '.join(sorted(names))
> + )
> + return ""
> +
> + peekstr = "[n_obs x n_vars]\n %d x %d" % tuple(tmp.shape)
> + peekstr += _makelayerstrings("obs", tmp.obs_count, tmp.obs_layers)
> + peekstr += _makelayerstrings("var", tmp.var_count, tmp.var_layers)
> + peekstr += _makelayerstrings("obsm", tmp.obsm_count, tmp.obsm_layers)
> + peekstr += _makelayerstrings("varm", tmp.varm_count, tmp.varm_layers)
> + peekstr += _makelayerstrings("uns", tmp.uns_count, tmp.uns_layers)
>
> Is X here missing?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.<
https://ci6.googleusercontent.com/proxy/BLD-6pSpXP-9-kMyPHTMgfzfue-Y7sZPk7xWTKvu__5drMa6yfliFO_dL1i5qr4VufSrDFVDKv5a6ujS7NCND1NuvbphLG51YS4gOrM_1B0luqmyZ8CQELZk4yasbO5W8HTfm_UW2XJAj-9N20NlvlGMU7oU0C8j-2v5MA7qJFOXRNGh_kVfUMYUdlWq4tTzpvIggPZabM_ruGuhkbiQAbXvFsk9-8rSDmDrksb74u9FusHV-A1L=s0-d-e1-ft#https://github.com/notifications/beacon/AE5PM6XIA5FKCRXHRLEH6BLR7UKTTA5CNFSM4PBO2FWKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGODOSOJHA.gif>

**Comment by bgruening** [comment]
@mtekman is this still WIP?

**Comment by mtekman** [comment]
nope, complete from my side -- the preview dataset capability is not implemented, but that is fine by me

**Comment by mvdbeek** [comment]
Thanks a lot @mtekman!

**Pull request merged**